File: 01-Programming-Languages/Ruby/Algorithms/DynamicProgramming/fibonacci.rb
=begin
Problem: Fibonacci Number
Description: Compute the nth Fibonacci number using dynamic programming (bottom-up).

Approach:
- Use iterative DP with constant space
- Base cases: fib(0) = 0, fib(1) = 1

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input:  n = 10
Output: 55
=end

def fib(n)
  return n if n <= 1
  prev2 = 0
  prev1 = 1
  (2..n).each do
    curr = prev1 + prev2
    prev2 = prev1
    prev1 = curr
  end
  prev1
end

(0..10).each { |i| puts "fib(#{i}) = #{fib(i)}" }

File: 01-Programming-Languages/Java/Algorithms/DynamicProgramming/Fibonacci.java
/*
 * Problem: Compute the nth Fibonacci number.
 * Approach: Dynamic programming with memoization (top-down).
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * Example: Input: n = 6 -> Output: 8
 */

public class Fibonacci {
    public static int fib(int n) {
        if (n <= 1) return n;
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    public static void main(String[] args) {
        int n = 6;
        System.out.println(fib(n));
    }
}
